platform/source/TARGET_CORTEX_M/TOOLCHAIN_IAR/except.S

```diff
@@ -42,114 +42,75 @@ FAULT_TYPE_USAGE_FAULT          EQU     0x40
-        STR     R2,[R1]
-        ADDS    R1,#8                   ; Add 8 here to capture LR next, we will capture SP later
-        LDR     R2,[R0,#20]             ; Capture LR
-        STR     R2,[R1]
-        ADDS    R1,#4
-        LDR     R2,[R0,#24]             ; Capture PC
-        STR     R2,[R1]
-        ADDS    R1,#4
-        LDR     R2,[R0,#28]             ; Capture xPSR
-        STR     R2,[R1]
-        ADDS    R1,#4
-        ; Adjust stack pointer to its original value and capture it
-        MOV     R3,R0
-        ADDS    R3,#0x20                ; Add 0x20 to get the SP value prior to exception
-        LDR     R6,=0x200
-        TST     R2,R6                   ; Check for if STK was aligned by checking bit-9 in xPSR value
-        BEQ     Fault_Handler_Continue1
-        ADDS    R3,#0x4
+        LDR     R7,=mbed_fault_context
+        LDR     R7,[R7]
+        LDMIA   R6!,{R0-R3}
+        STMIA   R7!,{R0-R3}             ; Capture R0..R3
+        POP     {R0-R3}
+        STMIA   R7!,{R0-R3}             ; Capture R4..R7
+        MOV     R0,R8
+        MOV     R1,R9
+        MOV     R2,R10
+        MOV     R3,R11
+        STMIA   R7!,{R0-R3}             ; Capture R8..R11
+        LDMIA   R6!,{R0,R2-R4}          ; Load R12,LR,PC,xPSR
+        ; Adjust stack pointer to its original value
+        MOVS    R1,R6
+        LSRS    R6,R4,#9                ; Check for if STK was aligned by checking bit-9 in xPSR value
+        BCC     Fault_Handler_Continue1
+        ADDS    R1,#0x4
 
 Fault_Handler_Continue1
-        MOV     R5,LR
-        LDR     R6,=0x10                ; Check for bit-4 to see if FP context was saved
-        TST     R5,R6
-        BNE     Fault_Handler_Continue2
-        ADDS    R3,#0x48                ; 16 FP regs + FPCSR + 1 Reserved
-
+        LSRS    R6,R5,#4                ; Check EXC_RETURN bit-4 to see if FP context was saved
+        BCS     Fault_Handler_Continue2
+        ADDS    R1,#0x48                ; 16 FP regs + FPCSR + 1 Reserved
 Fault_Handler_Continue2
-        MOV     R4,R1
-        SUBS    R4,#0x10                ; Set the location of SP in ctx
-        STR     R3,[R4]                 ; Capture the adjusted SP
-        MRS     R2,PSP                  ; Get PSP
-        STR     R2,[R1]
-        ADDS    R1,#4
-        MRS     R2,MSP                  ; Get MSP
-        STR     R2,[R1]
-        ADDS    R1,#4
-        MOV     R2,LR                   ; Get current LR(EXC_RETURN)
-        STR     R2,[R1]
-        ADDS    R1,#4
-        MRS     R2,CONTROL              ; Get CONTROL Reg
-        STR     R2,[R1]
+        STMIA   R7!,{R0-R4}             ; Capture R12,SP,LR,PC,xPSR
+        MRS     R0,PSP
+        MOV     R1,SP
+        MRS     R6,CONTROL
+        STMIA   R7!,{R0,R1,R5,R6}       ; Capture PSP,MSP,EXC_RETURN,CONTROL
         MOV     R0,R12
-        LDR     R3,=mbed_fault_context
-        LDR     R1,[R3]
-        BL      mbed_fault_handler
+        MOVS    R1,R7
+        SUBS    R1,#21*4
```

Comment:
what do we do here with substract, can you add a comment? 

Is this similar what we do on line 99 - ADS to R1 value 48

Reply:
That's rewinding R7 back to the start of the fault context structure. I'm about to change it to just a reload of the address constant anyway - that's simpler once I've killed an unnecessary pointer indirection.